exonum/src/blockchain/transaction.rs

```diff
@@ -0,0 +1,235 @@
+    /// #
+    /// use exonum::blockchain::Transaction;
+    /// use exonum::crypto::PublicKey;
+    /// use exonum::messages::Message;
+    /// # use exonum::blockchain::TransactionResult;
+    /// # use exonum::storage::Fork;
+    ///
+    /// message! {
+    ///     struct MyTransaction {
+    ///         // Transaction definition...
+    /// #       const TYPE = 1;
+    /// #       const ID = 1;
+    /// #       const SIZE = 32;
+    ///         field public_key:    &PublicKey    [0 => 32]
+    ///     }
+    /// }
+    ///
+    /// impl Transaction for MyTransaction {
+    ///     fn verify(&self) -> bool {
+    ///         self.verify_signature(self.public_key())
+    ///     }
+    ///
+    ///     // Other methods...
+    ///     // ...
+    /// #   fn execute(&self, _: &mut Fork) -> TransactionResult {
+    /// #       Ok(TransactionValue::Success)
+    /// #   }
+    /// }
+    /// # fn main() {}
+    fn verify(&self) -> bool;
+
+    /// Takes the current blockchain state via `fork` and can modify it if certain conditions
+    /// are met.
+    ///
+    /// # Notes
+    ///
+    /// - When programming `execute`, you should perform state-related checks before any changes
+    /// to the state and return early if these checks fail.
+    /// - If the execute method of a transaction raises a `panic`, the changes made by the
+    /// transactions are discarded, but the transaction itself is still considered committed.
+    /// - A transaction execution status (see `ExecutionStatus` for the details) is stored in the
+    /// blockchain and can be accessed through api.
+    ///
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # #[macro_use] extern crate exonum;
+    /// #
+    /// use exonum::blockchain::{Transaction, TransactionResult};
+    /// use exonum::crypto::PublicKey;
+    /// use exonum::storage::Fork;
+    ///
+    /// message! {
+    ///     struct MyTransaction {
+    ///         // Transaction definition...
+    /// #       const TYPE = 1;
+    /// #       const ID = 1;
+    /// #       const SIZE = 32;
+    /// #       field public_key:    &PublicKey    [0 => 32]
+    ///     }
+    /// }
+    ///
+    /// impl Transaction for MyTransaction {
+    ///     fn execute(&self, fork: &mut Fork) -> TransactionResult {
+    ///         // Read and/or write into storage.
+    ///         // ...
+    ///
+    ///         // Returns execution status.
+    ///         Ok(TransactionValue::Success)
+    ///     }
+    ///
+    ///     // Other methods...
+    ///     // ...
+    /// #   fn verify(&self) -> bool { true }
+    /// }
+    /// # fn main() {}
+    fn execute(&self, fork: &mut Fork) -> TransactionResult;
+}
+
+/// Result of successful transaction execution.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub enum TransactionValue {
+    /// Successful transaction execution.
+    Success,
+    /// User defined code. Can have different meanings for different transactions and services.
```

Comment:
Could you please specify an example of several successful statuses (probably, as an example section in the `TransactionValue` docs)? If I understand correctly,they could arise if the transaction logic contains branching execution paths; however, I'm having trouble understanding what `Success` means in this case.

Reply:
Right now I don't have a good example of several different successful statuses, but I suppose that they are needed because we are making a general-purpose (blockchain) framework.

As for `Success` - it is a simple predefined success status, same as `UnknownFailure`.